thirdparty/github.go

```diff
@@ -171,98 +198,62 @@ func githubShouldRetry(caller string) utility.HTTPRetryFunction {
-				"caller":    caller,
-				"retry_num": index,
-			})
-			return true
-		}
-
-		limit := parseGithubRateLimit(resp.Header)
-		if limit.Remaining == 0 {
-			return false
-		}
-
-		if resp.StatusCode == http.StatusNotFound {
-			logGitHubRateLimit(limit)
+		if config.retry404 && resp.StatusCode == http.StatusNotFound {
 			grip.Info(message.Fields{
 				"ticket":    GithubInvestigation,
-				"message":   fmt.Sprintf("hit %d in githubShouldRetryWith404s", http.StatusNotFound),
+				"message":   fmt.Sprintf("hit %d in githubShouldRetry", http.StatusNotFound),
 				"caller":    caller,
 				"retry_num": index,
 			})
 			return true
 		}
 
-		retryFn := githubShouldRetry(caller)
-		return retryFn(index, req, resp, err)
+		return false
 	}
 }
 
-func getGithubClientRetryWith404s(token, caller string) *http.Client {
-	grip.Info(message.Fields{
-		"ticket":  GithubInvestigation,
-		"message": "called getGithubClientRetryWith404s",
-		"caller":  caller,
-	})
-	return utility.GetOauth2CustomHTTPRetryableClient(
-		token,
-		githubShouldRetryWith404s(caller),
-		utility.RetryHTTPDelay(utility.RetryOptions{
-			MaxAttempts: numGithubAttempts,
-			MinDelay:    githubRetryMinDelay,
-		}),
-	)
-}
-
-func getGithubClient(token, caller string) *http.Client {
+// getGithubClient returns an *http.Client that will retry according to supplied retryConfig.
+// Also creates a span tracking the lifespan of the client.
+// Defer the returned function to release the client to the pool and close the span.
+func getGithubClient(ctx context.Context, token, caller string, config retryConfig, attributes []attribute.KeyValue) (context.Context, *http.Client, func()) {
 	grip.Info(message.Fields{
 		"ticket":  GithubInvestigation,
 		"message": "called getGithubClient",
 		"caller":  caller,
 	})
 
-	return utility.GetOAuth2HTTPClient(token)
-}
+	ctx, span := tracer.Start(ctx, caller, trace.WithAttributes(append(attributes, attribute.String(endpointAttribute, caller))...))
 
-// getGithubClientRetry will retry github operations if the error is temporary, resp is nil,
-// or we hit a bad gateway error.
-func getGithubClientRetry(token, caller string) *http.Client {
-	grip.Info(message.Fields{
-		"ticket":  GithubInvestigation,
-		"message": "called getGithubClientRetry",
-		"caller":  caller,
-	})
-	return utility.GetOauth2CustomHTTPRetryableClient(
-		token,
-		githubShouldRetry(caller),
-		utility.RetryHTTPDelay(utility.RetryOptions{
-			MaxAttempts: numGithubAttempts,
-			MinDelay:    githubRetryMinDelay,
-		}),
-	)
+	client := utility.GetHTTPClient()
+	originalTransport := client.Transport
+	client.Transport = otelhttp.NewTransport(client.Transport)
+
+	return ctx,
+		utility.SetupOauth2CustomHTTPRetryableClient(
+			token,
+			githubShouldRetry(caller, config),
+			utility.RetryHTTPDelay(utility.RetryOptions{
+				MaxAttempts: numGithubAttempts,
+				MinDelay:    githubRetryMinDelay,
+			}),
+			client,
+		), func() {
```

Comment:
😞 this was such a headache!
I realized that that function was unable to extract the transport from the otelhttp transport since it's [private](https://github.com/open-telemetry/opentelemetry-go-contrib/blob/e94326455e25cede6b4381d3f90efae18f71b29a/instrumentation/net/http/otelhttp/transport.go#L33). So the only way to reset the transport without making a whole new one is to hold on to the client we give the otelhttp transport and reinstate it before we call `PutHTTPClient`.